### PR TITLE
✨  Add `remediationsAllowed` field to MHC status

### LINKED ISSUE #3371: Add `allowedRemediations` field to MHC Status
**User Story**

As a user I would like to know how many Machines an MHC thinks it can disrupt before short circuiting to increase confidence that the MHC will not break my cluster when it needs to make remediations.

**Detailed Description**

I think we should add a new field to the MHC status called `allowedRemediations` that will give users some insight into how the MHC is going to make decisions when it comes across an unhealthy machine.

We already calculate the maximum number of unhealthy nodes in `isAllowedRemediation`, I would expect the value of this field to be `maxUnhealthy - unhealthy`, except when unhealthy is greater than maxUnhealthy in which case we would set it to 0, implying that short circuiting of the remediation will happen for any currently unhealthy, unremediated machines
https://github.com/kubernetes-sigs/cluster-api/blob/b9bcc8c55bea0cad1b22cd872162d06c8687befe/controllers/machinehealthcheck_controller.go#L421-L434

While this can be calculated from entries in the status already, I think it would be clearer for users if it were on the status for them (this would also prevent mismatching of expectations of the `maxUnhealthy` field since it would give a definitive answer for how many machines the MHC thinks it's allowed to remove)

eg:
```
spec:
  maxUnhealthy: 40%
status:
  expectedMachines: 8
  currentHealthy: 6
  allowedRemediations: 1
```

**Anything else you would like to add:**

Happy to raise a PR for this, but thought I should seek feedback first on this idea

/kind feature

### PR DESCRIPTION
**Pull request by JoelSpeed**
**What this PR does / why we need it**:
Adds a new field to the MHC status called `remediationsAllowed` as proposed in #3371 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3371

### CODE REVIEW

**Comment by JoelSpeed** [comment]
> but just having the number of 0 might misinform and indicate that it's broken.

I don't know if having it as a 0 indicates it's broken necessarily, if it's 0 for a long time yes, but if it's 0 that just indicates you've used up all your allowed remediations for now and that no further remediations will be allowed. I would expect it to hit 0 quite regularly during normal use, just as you would see 0 on `allowedDisruptions` on a PDB quite often in normal use

File: api/v1alpha3/machinehealthcheck_types.go
Comment (vincepri) [inline api/v1alpha3/machinehealthcheck_types.go:87]:
```suggestion
	// AllowedRemediations is the number of further remediations allowed by this machine health check before
```

File: api/v1alpha3/machinehealthcheck_types.go
Comment (vincepri) [inline api/v1alpha3/machinehealthcheck_types.go:90]: Any other name that'd be suitable here? I find "Allowed" a little misleading, but I don't have a good suggestion

Comment (JoelSpeed) [reply api/v1alpha3/machinehealthcheck_types.go:90]: Hmm, I'll have to have a think on that one, welcome to suggestions. I had taken the inspiration from the `allowedDisruptions` field on a PDB, I feel like this represents a very similar concept

Comment (benmoss) [reply api/v1alpha3/machinehealthcheck_types.go:90]: I think allowed makes sense, especially after seeing it in this context (from #3013)
```yaml
spec:
  maxUnhealthy: 40%
status:
  expectedMachines: 8
  currentHealthy: 6
  allowedRemediations: 1
```

**Comment by benmoss** [comment]
I think from discussion on #3371 this is WIP and there's a plan to add conditions to this

**Comment by JoelSpeed** [comment]
I've added a condition to the PR, though I haven't updated the tests since I think I'm going to have to get quite involved to write a Matcher for the conditions. Do the extra couple of commits look right based on what we were discussing? CC @vincepri @michaelgugino

**Comment by vincepri** [comment]
Latest code changes look good, it looks super clean!

cc @fabriziopandini

**Comment by JoelSpeed** [comment]
Just FYI, I started working on updating the tests, but didn't manage to finish, so will push Monday. But I added a condition matcher for gomega like we have for the MHC status.
I am letting it ignore the timestamp, but match everything else, does that seem right?

**Comment by JoelSpeed** [comment]
This should now pass the tests.

I have kept both the condition and allowedRemediations integer field for now as I wasn't sure what the intention was to do there, personally I can see value to both so would prefer to keep both

File: controllers/machinehealthcheck_controller.go
Comment (michaelgugino) [inline controllers/machinehealthcheck_controller.go:232]: I think this comment text should be what we're using for the message text.  It's clearer than the existing message.

File: api/v1alpha3/machinehealthcheck_types.go
Comment (michaelgugino) [reply api/v1alpha3/machinehealthcheck_types.go:90]: I think maybe "CurrentRemediationBudget" might make more sense.  I think this better confers to the user what the user has left. "AllowedRemediations" makes it look like we're over the limit already.

Overall, I'm not sure I like this field.  I think maybe a boolean of "RemediationsAllowed" would be clearer here, even if redundant to the status object.

File: controllers/machinehealthcheck_controller.go
Comment (JoelSpeed) [reply controllers/machinehealthcheck_controller.go:232]: Would you include the values still or omit those? 
Eg:
```
Remediation not allowed, the number of not started or unhealthy machines exceeds maxUnhealthy (total: 4, unhealthy: 3, maxUnhealthy: 40%)
```

File: api/v1alpha3/machinehealthcheck_types.go
Comment (JoelSpeed) [reply api/v1alpha3/machinehealthcheck_types.go:90]: Personally I think it's useful to have a numeric field in the status for the number of further remediations that are allowed as it will give users an idea of how many machines the controller thinks can go further unhealthy, eg, on a percentage based system where rounding is involved, it shows which way the controller rounds. It may also be helpful tracking down bugs in the future, eg where `"10"` was treated as `"10%"` because of an error in the intorstring implementation, you could immediately see that the number the controller is interpreting is wrong.

I would prefer `RemediationBudget` to `CurrentRemediationBudget` as I feel the word `Current` is a bit superfluous, status fields are implicitly the "current" state

File: controllers/machinehealthcheck_controller.go
Comment (michaelgugino) [reply controllers/machinehealthcheck_controller.go:232]: Yeah, the totals are fine, I'm just talking about the text part.

File: api/v1alpha3/condition_consts.go
Comment (CecileRobertMichon) [inline api/v1alpha3/condition_consts.go:104]:
```suggestion
	RemediationAllowedCondition ConditionType = "RemediationAllowed"
```

File: controllers/machinehealthcheck_controller.go
Comment (CecileRobertMichon) [inline controllers/machinehealthcheck_controller.go:469]: slight preference for returning an error here if somehow that value isn't set rather than silently defaulting in the controller

File: controllers/machinehealthcheck_controller.go
Comment (CecileRobertMichon) [inline controllers/machinehealthcheck_controller.go:455]: this isn't from this PR but shouldn't remediation only be allowed if unhealthy is strictly less than maxUnhealthy? If unhealthy == maxUnhealthy and further remediation is attempted, wouldn't that risk n + 1 machines to become unhealthy ?

File: api/v1alpha3/machinehealthcheck_types.go
Comment (benmoss) [reply api/v1alpha3/machinehealthcheck_types.go:87]: Agreed, but to be fair the other comments on this struct are pretty inconsistent

File: controllers/machinehealthcheck_controller.go
Comment (benmoss) [reply controllers/machinehealthcheck_controller.go:455]: This [isn't the cleanest wording](https://github.com/kubernetes-sigs/cluster-api/blob/d4101d79f33001db88e2ca48dc485ed6ccaa8727/api/v1alpha3/machinehealthcheck_types.go#L43-L44) in the English language but

> Any further remediation is only allowed if at most "MaxUnhealthy" machines selected by "selector" are not healthy.

I would say this means "Remediation is allowed if `MaxUnhealthy` machines are unhealthy."

**Comment by vincepri** [comment]
/milestone v0.3.10

Moving the milestone on this one, unless we can finish it by tomorrow/Friday

**Comment by JoelSpeed** [comment]
I'm on vacation so won't have time to fix until the week of the 7th of September

File: controllers/machinehealthcheck_controller.go
Comment (JoelSpeed) [reply controllers/machinehealthcheck_controller.go:455]: I agree the wording isn't the best, but I think Ben hit the nail on the head here, I think the intention of this field is that if `maxUnhealthy: 3` and the current number of unhealthy machines is 3, those three machines can be remediated, however, if 4 were unhealthy, none of the unhealthy machines would be remediated.

This PR just refactors the logic, nothing should have changed from the behaviour here (or at least that was my intention)

File: controllers/machinehealthcheck_controller.go
Comment (JoelSpeed) [reply controllers/machinehealthcheck_controller.go:469]: That would change the existing logic, currently we just return allowed if maxUnhealthy is not set, so this 100% is acting as a proxy for that allowance.

I could return an error and then check if it's a particular "not set" error and behave appropriately then, WDYT?
```
maxUnhealthy, err := getMaxUnhealthy(mhc)
if err == MaxUnhealthyNotSetErr {
  return true, nil 
} else if err != nil {
  return false, err
}
```

I don't know if this really adds that much though, perhaps a better explanation of why we default here would be better?

**Comment by JoelSpeed** [comment]
I've rebased this to fix the conflicts and should have addressed or responded to all comments, PTAL folks

**Comment by vincepri** [comment]
/retest

@CecileRobertMichon @fabriziopandini Do you all have time to do another review?

File: api/v1alpha3/machinehealthcheck_types.go
Comment (CecileRobertMichon) [reply api/v1alpha3/machinehealthcheck_types.go:90]: `remediationsAllowed` makes slightly more sense to me, especially since that is a closer match to `disruptionsAllowed` https://kubernetes.io/docs/tasks/run-application/configure-pdb/#check-the-status-of-the-pdb

**Comment by vincepri** [comment]
@JoelSpeed Are you still interested in getting this change in before release?

**Comment by JoelSpeed** [comment]
@vincepri I should be able to get some time to look at this over the next couple of days, when's the deadline for the next release?

**Comment by vincepri** [comment]
October 1st

File: controllers/machinehealthcheck_controller.go
Comment (JoelSpeed) [reply controllers/machinehealthcheck_controller.go]: Updated as per your suggestion. I've added a TODO comment above the code and will follow up on this with a PR to do that change. I don't think it's too major but I'd prefer to keep any discussion about a change of logic to a separate PR so that we don't hold up this UX change

**Comment by JoelSpeed** [comment]
For those following this, I believe I have resolved all of the comments on this now, and the many conflicts in the tests.

I think this is ready to go so please take another look over

**Review by CecileRobertMichon — APPROVED** [review]
lgtm

@JoelSpeed can you update the commit message / PR description to change `allowedRemediations` to `remediationsAllowed`

**Comment by JoelSpeed** [comment]
@CecileRobertMichon Should I squash everything down to a single commit as I do that? Wondering if it's worth keeping the conditions status matcher stuff as a separate commit 🤔

**Comment by CecileRobertMichon** [comment]
Up to you, one commit for the conditions matcher and one for the rest sgtm

**Comment by JoelSpeed** [comment]
@CecileRobertMichon Thanks, all updated

**Comment by wfernandes** [comment]
Taking a look 👀

File: api/v1alpha3/condition_consts.go
Comment (wfernandes) [inline api/v1alpha3/condition_consts.go:135]:
```suggestion
	// allowed to remediate any Machines or whether it is blocked from remediating any further.
```

File: controllers/machinehealthcheck_controller_test.go
Comment (wfernandes) [inline controllers/machinehealthcheck_controller_test.go:1554]: minor nit: 
This was a little confusing to read since this is the actual total machines counted by the MHC. Maybe instead of `expectedMachines` we can say `actualExpectedMachines` or `actualMachineCount`.

**Review by wfernandes — COMMENTED** [review]
Some really minor nits. But other than these,
/lgtm

File: controllers/machinehealthcheck_controller.go
Comment (vincepri) [inline controllers/machinehealthcheck_controller.go:234]:
```suggestion
		message := fmt.Sprintf("Remediation is not allowed, the number of not started or unhealthy machines exceeds maxUnhealthy (total: %v, unhealthy: %v, maxUnhealthy: %v)",
			totalTargets,
			m.Status.CurrentHealthy,
			m.Spec.MaxUnhealthy,
		)
```

File: controllers/machinehealthcheck_controller.go
Comment (vincepri) [inline controllers/machinehealthcheck_controller.go:481]:
```suggestion
	if mhc.Spec.MaxUnhealthy == nil {
		return 0, errors.New("spec.maxUnhealthy must be set")
```

Is this something that's validated or defaulted?

File: controllers/machinehealthcheck_controller.go
Comment (vincepri) [inline controllers/machinehealthcheck_controller.go:460]: This is also checked on line 462, should it return true or false when it's nil?

**Comment by JoelSpeed** [comment]
@wfernandes @vincepri I applied your suggestions and responded to questions, please take another look when you have a moment

**Comment by vincepri** [comment]
I don't see the answers to the question from my review

**Comment by vincepri** [comment]
@JoelSpeed Given the changes in this PR are extensive to merge 1 day before release, and without enough time for CI to do multiple runs for a few days, is it ok to change the milestone to v0.3.11?

**Comment by JoelSpeed** [comment]
I'm on my phone so can't reply fully. I can see my comments which I thought I had sent, but they are marked as pending and I can't seem to submit now, not sure what happened there 🤔

Anyway, I'm not in a rush to merge so happy to defer to 0.3.11 if you'd prefer. I think we have reached consensus on the design so that unblocks my implementation of the same feature in OpenShift

**Comment by vincepri** [comment]
/milestone v0.3.11

Thank you!

**Comment by vincepri** [comment]
Changed the base branch to be release-0.3, we'll have to forward-port these changes to v0.4 later

**Comment by JoelSpeed** [comment]
/retest

@vincepri Not handled a PR into a previous release before on this project, is there anything in particular I need to be doing now? I think I may need another review from you if you have a a moment

**Comment by vincepri** [comment]
cc @CecileRobertMichon @wfernandes as well, for another pass

**Comment by wfernandes** [comment]
I'll probably get to it tomorrow, if that works.

File: controllers/machinehealthcheck_controller.go
Comment (wfernandes) [inline controllers/machinehealthcheck_controller.go:467]: This may be a nit, but I'd be very clear with this comment. I say this because the phrase "short circuit" may require additional "thinking" especially for those who are not fluent in English or Electronics 😄 .
```suggestion
	// Remediation is not allowed if unhealthy is above maxUnhealthy
```

File: controllers/machinehealthcheck_controller_test.go
Comment (wfernandes) [inline controllers/machinehealthcheck_controller_test.go:336]: Yeah, so to my previous point, in this case unhealthy is 2 and not 1.

**Review by wfernandes — COMMENTED** [review]
Thanks for this PR @JoelSpeed!

**Comment by JoelSpeed** [comment]
@wfernandes Thanks for the review, think I've addressed all of your suggestions now

**Comment by CecileRobertMichon** [comment]
In terms of backporting, given that this PR is targeting release-0.3 branch, we should make sure to open a new PR on the main branch with the ~same changes

@JoelSpeed are you taking care of that PR?

/approve

**Comment by JoelSpeed** [comment]
Will forward port on Monday!

**Pull request merged**
